ce/api/data.py

```diff
@@ -72,40 +72,79 @@ def data(sesh, model, emission, time, area, variable, timescale='other',
-
-    results = sesh.query(Run)\
-            .join(Model, Emission, DataFile, DataFileVariable, TimeSet)\
-            .join(EnsembleDataFileVariables, Ensemble)\
-            .filter(DataFileVariable.netcdf_variable_name == variable)\
-            .filter(Emission.short_name == emission)\
-            .filter(Model.short_name == model)\
-            .filter(TimeSet.time_resolution == timescale)\
-            .filter(Ensemble.name == ensemble_name).all()
-
-    if not results:
-        return {}
-
-    def getdata(file_, time_idx):
-        with open_nc(file_.filename) as nc:
-            a = get_array(nc, file_.filename, time_idx, area, variable)
+        raise Exception('time parameter "{}" not convertable to an integer.'
+                        .format(time))
+
+    def get_spatially_averaged_data(data_file, time_idx):
+        """
+        From the NetCDF data file pointed at by `data_file`,
+        get the spatial average over the area specified by `area`
+        of the data for variable `variable`
+        at time index `time_idx`.
+
+        :param data_file (modelmeta.DataFile): source data file
+        :param time_idx (int): index of time of interest
+        :return: float
+        """
+        with open_nc(data_file.filename) as nc:
+            a = get_array(nc, data_file.filename, time_idx, area, variable)
         return np.asscalar(np.mean(a))
 
-    def get_timeval(timeset, idx):
+    def get_time_value(timeset, time_idx):
+        """
+        Get the time value associated with time index `time_idx`
+        from the time set `timeset`.
+
+        :param timeset (modelmeta.TimeSet): time set
+        :param time_idx (int): index of time of interest
+        :return: (str)
+        """
         for time in timeset.times:
-            if time.time_idx == idx:
+            if time.time_idx == time_idx:
                 return time.timestep
-        raise Exception('Timeset has no time with index value {}'.format(idx))
-
-    return {
-        run.name: {
-            'data': {
-                get_timeval(file_.timeset, time).strftime('%Y-%m-%dT%H:%M:%SZ'):
-                        getdata(file_, time) for file_ in get_files_from_run_variable(run, variable)
-            },
-            'units': get_units_from_run_object(run, variable)
-        } for run in results
-    }
+        raise Exception('Timeset has no time with index value {}'
+                        .format(time_idx))
+
+    query = (
+        sesh.query(DataFileVariable)
+        .filter(DataFileVariable.netcdf_variable_name == variable)
+
+        .join(DataFileVariable.file)
+        .join(DataFile.run)
+
+        .join(Run.model)
+        .filter(Model.short_name == model)
+
+        .join(Run.emission)
+        .filter(Emission.short_name == emission)
+
+        .join(DataFile.timeset)
+        .filter(TimeSet.time_resolution == timescale)
+
+        .filter(DataFileVariable.ensembles.any(Ensemble.name == ensemble_name))
+    )
```

Comment:
I grouped related joins and filters, but this may not seem to others, as it does to me, a clearer way to organize the query.